dss-collection-nft/contracts/DSSCollection.cdc

```diff
@@ -0,0 +1,695 @@
+            }
+
+            let item = DSSCollection.getItem(id: id)
+            self.items.append(item)
+
+            emit ItemAddedToSlot(
+                 id: item.id,
+                 slotID: self.id,
+                 collectionGroupID: self.collectionGroupID
+             )
+        }
+
+        init (
+            collectionGroupID: UInt64,
+            logicalOperator: String,
+            typeName: String
+        ) {
+            pre {
+                DSSCollection.CollectionGroupData(
+                    id: collectionGroupID
+                ).open == true: "collection group not open"
+            }
+
+            self.id = self.uuid
+            self.collectionGroupID = collectionGroupID
+            self.logicalOperator = logicalOperator
+            self.typeName = typeName
+            self.items = []
+
+            emit SlotCreated(
+                id: self.id,
+                collectionGroupID: self.collectionGroupID,
+                logicalOperator: self.logicalOperator,
+                typeName: self.typeName
+            )
+        }
+    }
+
+    // A public struct to access CollectionGroup data
+    //
+    pub struct CollectionGroupData {
+        pub let id: UInt64
+        pub let name: String
+        pub let description: String
+        pub let typeName: String
+        pub let open: Bool
+        pub let startTime: UFix64?
+        pub let endTime: UFix64?
+        pub let timeBound: Bool
+
+        init (id: UInt64) {
+            if let collectionGroup = &DSSCollection.collectionGroupByID[id] as &DSSCollection.CollectionGroup? {
+                self.id = collectionGroup.id
+                self.name = collectionGroup.name
+                self.description = collectionGroup.description
+                self.typeName = collectionGroup.typeName
+                self.open = collectionGroup.open
+                self.startTime = collectionGroup.startTime
+                self.endTime = collectionGroup.endTime
+                self.timeBound = collectionGroup.timeBound
+            } else {
+                panic("collectionGroup does not exist")
+            }
+        }
+    }
+
+    // A top-level CollectionGroup with a unique ID and name
+    //
+    pub resource CollectionGroup {
+        pub let id: UInt64
+        pub let name: String
+        pub let description: String
+        pub let typeName: String
+        pub var open: Bool
+        pub let startTime: UFix64?
+        pub let endTime: UFix64?
+        pub let timeBound: Bool
+        pub var numMinted: UInt64
+
+        // Close this collection group
+        //
+        access(contract) fun close() {
+            pre {
+                self.open == true: "not open"
+            }
+
+            self.open = false
+
+            emit CollectionGroupClosed(id: self.id)
+        }
+
+        // Mint a DSSCollection NFT in this group
+        //
+        pub fun mint(completedBy: String, level: UInt8): @DSSCollection.NFT {
```

Comment:
Can you please give me more information about `completedBy` param it is hard to understand how it is useful.

Reply:
It's the wallet address of who did the challenge.  I thought it as a timestamp first time I read it, but couldn't think of a better name.  Maybe completerAddress, or completionAddress?  naming can be hard...